larq_compute_engine/core/bconv2d_output_transform.h

```diff
@@ -21,93 +21,128 @@ enum class OutputTransformDetails {
+ * clamp values and fusing (part of) the back-transformation into the
+ * multiplier/bias:
+ *                 float_cast(clamp'(accum << 1)) * mul' + bias'
+ *
+ * Note that the left shift cannot be fused, because the left-shifted
+ * accumulator value will either be always odd or always even, and we support
+ * clamping to arbitrary integers.
+ *
+ * The parameters of the output transform are therefore two int32 clamp values
+ * and float pointers to a multiplier and a bias.
+ */
+template <>
+struct OutputTransform<float, OutputTransformDetails::Default> {
+  std::int32_t clamp_min = std::numeric_limits<std::int32_t>::lowest();
+  std::int32_t clamp_max = std::numeric_limits<std::int32_t>::max();
+  const float* multiplier = nullptr;
+  const float* bias = nullptr;
+
+  float Run(const std::int32_t accum, int out_channel) const {
+    TFLITE_DCHECK(multiplier != nullptr);
+    TFLITE_DCHECK(bias != nullptr);
+    std::int32_t x = accum << 1;
+    x = std::max<std::int32_t>(std::min<std::int32_t>(x, clamp_max), clamp_min);
+    return static_cast<float>(x) * multiplier[out_channel] + bias[out_channel];
   }
 };
 
-// Output transformation for float kernels
-template <typename AccumScalar>
-struct OutputTransform<AccumScalar, float, OutputTransformDetails::Default>
-    : OutputTransformBase<AccumScalar> {
-  const float* post_activation_multiplier = nullptr;
-  const float* post_activation_bias = nullptr;
-
-  float Run(const AccumScalar accum, int out_channel) const {
-    TF_LITE_ASSERT(post_activation_multiplier != nullptr);
-    TF_LITE_ASSERT(post_activation_bias != nullptr);
-    // Post multiply and add are done in float
-    float x = static_cast<float>(this->RunBase(accum));
-    x *= post_activation_multiplier[out_channel];
-    x += post_activation_bias[out_channel];
-    return x;
+/*
+ * -----------
+ * Int8 output
+ * -----------
+ *
+ * For int8 output, there are two additional conceptual steps:
+ *   4. Multiply by the reciprocal of the int8 output scale, and add the int8
+ *      zero-point.
+ *   5. Round to an integer and clamp to the int8 range.
+ *
+ * We can fuse step (4) into the existing multiplier/bias, and so the int8
+ * output transform is:
+ *           int8_cast(float_cast(clamp'(accum << 1)) * mul + bias)
+ *
+ * Thus, the int8 output transform parameters are the same as in the float case.
+ */
+template <>
+struct OutputTransform<std::int8_t, OutputTransformDetails::Default> {
+  std::int32_t clamp_min = std::numeric_limits<std::int32_t>::lowest();
+  std::int32_t clamp_max = std::numeric_limits<std::int32_t>::max();
+  const float* multiplier = nullptr;
+  const float* bias = nullptr;
+
+  std::int8_t Run(const std::int32_t accum, int out_channel) const {
+    TFLITE_DCHECK(multiplier != nullptr);
+    TFLITE_DCHECK(bias != nullptr);
+    // Clamping is done in int32
+    std::int32_t x = accum << 1;
+    x = std::max<std::int32_t>(std::min<std::int32_t>(x, clamp_max), clamp_min);
```

Comment:
What do the clamps do? I assume the compiler will eliminate them as they statically resolve to true.  If they weren't there you could absorb the shift left into the multiplier array.

Reply:
The `clamp_min` , `clamp_max` values depend on the `"fused_activation_function"` attribute. They effectively take care of a ReLu. We could write two versions of the `OutputTransform`, with and without activation function. The latter could then indeed absorb the shift into the multiplier, and skip the clamp.

Reply:
I see, that makes sense, thanks.